### PR TITLE
Mobile Menu Functionality

### LINKED ISSUE #298: Build mobile menu JS functionality
### Build JavaScript to display and interface with Navigation API. for the Mobile Menu on Cancer.gov

The mobile menu is a new component specific to NCI.  This ticket includes the JavaScript required to produce the mobile navigation from the set API listed bellow. This includes the fetching, consuming and display of the data as HTML DOM elements using the Navigation API in the ticket listed below. 

> **ESTIMATE** 5
Re-estimated on 5/16: 3

###related tickets: https://github.com/NCIOCPL/cgov-digital-platform/issues/3278 & https://github.com/NCIOCPL/cgov-digital-platform/issues/3243

### USWDS Component Equivalent
The mobile menu is a NCI unique component that has no USWDS equivalent. The design pattern breaks from their uswds header once the mobile MENU button is clicked.  

### Behaviors
*Click [Menu] Button to open Mobile Menu
*Click off or [X] Button to close Mobile Menu
*Click [Section] Parent Link to retrieve json data based on API/ID
*Display current json data
*Click to go [Page Link] to go to url
*Optional tracking callBack on clicks

### API information
input: ```taxonomy_term```
return:
```
{
  id: number, // The term id of the taxonomy node - this should be a unique id
  langcode: 'string', // The 2 char language code
  label: 'string', // This is the field_navigation_label field, or if empty, the name field
  weight: number, // The order of the term
  path: 'string', // The alias of the field_landing_page target
  navigation_display_options: [ ... parameters ... ], // the array of parameters
  is_section_nav_root: boolean, // true if this Site Section is a section nav root
  is_main_nav_root: boolean, // true if this Site Section is a main nav root
  children: [ ... NavItem ... ] // 0 or more NavItem objects
}
```

### InVision: 
  * DSM link: N/A
  * Desktop mockup: N/A
  * Tablet mockup: (https://publicis.invisionapp.com/share/4G11TDSW58SE#/screens/459263421)
  * Mobile mockup:(https://publicis.invisionapp.com/share/4G11TDSW58SE#/screens/459263421)

### Acceptance Criteria

```gherkin
Scenario: User is on a page and opens and closes the mobile menu 
  Given they are on a screen 1023px and smaller
  When the menu button is clicked
  Then the mobile menu appears, sliding in from the left 
    And covers 80% of the screen
    And has an overlay of the content behind it at rgba(0, 0, 0, 0.2);
    And the current page has an indicator (blue line and bolded blue text)
  When the "x" button is clicked
  Then the mobile menu slides left and disappears
```

### Prerequisites
- *dependency* Mock data will be used till API is ready and able to talk with JS
https://github.com/NCIOCPL/cgov-digital-platform/issues/3243

Link to Technical Design: https://github.com/NCIOCPL/ncids/wiki/Technical-NCI-Header-Design#mobile-menu-adapters

### Tasks
- [ ] New NCIDS-JS file for MobileNavigation
  - [ ] Async call to API
  - [ ] Processing / Sorting for Display based on current page
  - [ ] Menu Open/Close
     - [ ] Reset Navigation State on Close 
  - [ ] Section Nav Toggle to show/hide children
     - [ ] Reset Navigation State on Close        
  - [ ] Class for Display of Navigation Node
     - [ ] States for Default/Active
     - [ ] Section Navigation 
     - [ ] Navigation Link

### PR DESCRIPTION
**Pull request by karlikpj**
(#298) Mobile Menu Functionality

  ## Viewing Information
  * [The documentation site](https://designsystem-dev.cancer.gov/pr-456/)
  * [The storybook site](https://designsystem-dev.cancer.gov/pr-456/storybook/)
  * [The example site](https://designsystem-dev.cancer.gov/pr-456/example-site/)

Closes #298 .

### CODE REVIEW

**Comment by bryanpizzillo** [comment]
@karlikpj can you fix the commit message here to be:
(#298) Mobile Menu Functionality

Thanks

File: packages/ncids-js/src/components/nci-header/utils/mobile-menu/mobile-menu.ts [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+		} else {
+			this.activeMenu = true;
+			this.mobileNav.classList.add('active');
+			this.mobileOverlay.classList.toggle('active');
+			// pull data
+			const path = this.pageUrl;
+			this.lastClicked = path;
+			this.lastClicked = path;
+
+			const results = await this.adaptor.getInitialMenuId(path);
```
Comment (bryanpizzillo) [inline packages/ncids-js/src/components/nci-header/utils/mobile-menu/mobile-menu.ts:194]: This I think should be:
`const results = await this.adaptor.getNavigationLevel(this.adaptor.getInitialMenuId())`

**Comment by olitharp-nci** [comment]
Closes #298

**Comment by olitharp-nci** [comment]
Screen Shot 2022-07-11 at 10 12 28 AM

The mobile menu button color should not change
https://publicis.invisionapp.com/console/Header-&-Footer-Wireframes-cl17806xjx3ho01x42xix33gc/cl178079m77pi01uqi3oen68y/play

File: testing/ncids-js-testing/src/MockMobileMenuAdaptor.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+		}
+		const path = id === '/' ? '/root-menu' : id;
+		let returnData;
+		try {
+			const response = await fetch(`${rootPath}/data/mobile-menu${path}.json`);
+			const data = response.json();
+			returnData = data;
+			return data;
+		} catch (err: unknown) {
+			console.log(err);
```
Comment (olitharp-nci) [inline testing/ncids-js-testing/src/MockMobileMenuAdaptor.ts:30]: console.error(err);

File: packages/ncids-js/src/components/nci-header/utils/mobile-menu/mobile-menu.ts [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+	 *
+	 * @returns {HTMLElement}
+	 * @private
+	 */
+	private makeBackNode(item: MobileMenuItem): HTMLElement {
+		let label;
+		let path;
+		const lang = this.langCode;
+		if (this.sectionParent === undefined) {
+			path = '/';
```
Comment (olitharp-nci) [inline packages/ncids-js/src/components/nci-header/utils/mobile-menu/mobile-menu.ts:300]: Looks like there is missing code coverage for the first part of the if statement. Second half looks covered?

**Review by olitharp-nci — CHANGES_REQUESTED** [review]

File: packages/ncids-js/src/components/nci-header/__tests__/mobile-menu/mobile-menu-adaptor.mock.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	constructor(useUrlForNavigationId: boolean) {
+		this.useUrlForNavigationId = useUrlForNavigationId;
+	}
+
+	async getInitialMenuId(): Promise<string | number> {
+		return '/root-menu';
+	}
+
+	async getNavigationLevel(id: string): Promise<MobileMenuData> {
+		console.log(id);
```
Comment (olitharp-nci) [inline packages/ncids-js/src/components/nci-header/__tests__/mobile-menu/mobile-menu-adaptor.mock.ts:18]: remove

**Review by olitharp-nci — CHANGES_REQUESTED** [review]

File: testing/ncids-js-testing/public/index.html [hunk 5]
```diff
@@ -113,7 +113,6 @@ <h3 class="usa-alert__heading">COVID-19 resources.</h3>
 					</ul>
 				</div>
 			</nav>
-			<div class="nci-header-mobilenav__overlay"></div>
```
Comment (arcepaul) [inline testing/ncids-js-testing/public/index.html:116]: need to update everywhere

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
The text of the Search button does not submit the search form. The Search button text should be switched from a `<label>` element to a `<span>`.
```
 <button class="usa-button nci-header-search__search-button" type="submit" aria-label="Search">
          <span class="nci-header-search__label" aria-hidden="true">Search</span>
```

**Review by olitharp-nci — APPROVED** [review]

**Comment by bennettcc** [comment]
This PR has been passed to @mworrest and @meglake for review

**Review by mworrest — APPROVED** [review]
I can only interact with it so much on the example site, but the various states captured in storybook seem to cover the use cases.

**Pull request merged**